bindings/rust/s2n-tls/src/testing.rs

```diff
@@ -325,3 +335,205 @@ pub fn poll_tls_pair_result(pair: &mut Pair<Harness, Harness>) -> Result<()> {
         }
     }
 }
+
+type LocalDataBuffer = RefCell<VecDeque<u8>>;
+
+/// TestPair is a testing utility used to easily test handshakes and send data.
+///
+/// SAFETY: if the server or client connection is moved outside of the struct, IO
+/// is not safe to perform. The connections use pointers to data buffers owned by
+/// the Harness. If the Harness goes out of scope, the data buffers will be dropped
+/// and the pointers will be invalid.
+///
+/// The most common usecase is handshaking a simple config.
+/// ```ignore
+/// // given some config
+/// let config = build_config(&crate::security::DEFAULT_TLS13).unwrap();
+/// // create a pair (client + server) with uses that config
+/// let mut pair = TestPair::from_config(&config);
+/// // assert a successful handshake
+/// assert!(pair.handshake().is_ok());
+/// // we can also do IO using the poll_* functions
+/// // this data is sent using the shared data buffers owned by the harness
+/// assert!(pair.server.poll_send(&[3, 1, 4]).is_ready());
+/// let mut buffer = [0; 3];
+/// assert!(pair.client.poll_recv(&mut buffer).is_ready());
+/// assert_eq!([3, 1, 4], buffer);
+/// ```
+//
+// The doctest is manually checked in the manual_doctest test. This is because none
+// of the testing utilities are publicly exported and therefore can't be referenced
+// in a doc comment.
+//
+// We allow dead_code, because otherwise the compiler complains about the tx_streams
+// never being read. This is because it can't reason through the pointers that were
+// passed into the s2n-tls connection io contexts.
+#[allow(dead_code)]
+pub struct TestPair {
+    pub server: connection::Connection,
+    pub client: connection::Connection,
+
+    // Pin: since we are dereferencing this pointer (because it is passed as the send/recv ctx)
+    // we need to ensure that the pointer remains in the same place
+    // Box: A Vec (or VecDeque) may be moved or reallocated, so we need another layer of
+    // indirection to have a stable (pinned) reference
+    /// a data buffer that the server writes to and the client reads from
+    server_tx_stream: Pin<Box<LocalDataBuffer>>,
+    /// a data buffer that the client writes to and the server reads from
+    client_tx_stream: Pin<Box<LocalDataBuffer>>,
+}
+
+impl TestPair {
+    /// utility method to test simple handshakes
+    ///
+    /// Create a client and server from the associated `config`, and try to complete
+    /// a TLS handshake. The result of the handshake is returned.
+    pub fn oneshot_handshake(config: &config::Config) -> Result<(), error::Error> {
```

Comment:
Updated to `handshake_with_config`. 

`TestPair::handshake` and `TestPair.handshake` actually aren't allowed, you get an error about duplicate definitions.

Reply:
No, I meant make this a stand-alone function instead of a method on TestPair. Like you move it out of the impl. So just handshake and TestPair::handshake.

Reply:
Hmm, I'm generally not a fan of that approach. I worry that having it sit in a separate namespace makes it significantly less discoverable, and therefore people are less likely to use it.

Having all of our "local IO handshake" functions grouped under a compact namespace makes it significantly easier to peruse all of the available functionality.

Reply:
I think I'd feel more comfortable with that if we cleaned up our import path hygiene in our tests. I agree that `testing::handshake_with_config` would make sense as a complement to `testing::TestPair::handshake`, but right now so many of test modules glob import testing through `testing::*`, and I worry that `handshake_with_config()` sitting all alone by itself is going to be unpleasant.

It's the same way that I like `fs::read_to_string()`, but dislike `read_to_string()`.

Reply:
My long term goal is that unit tests modules should not be children of the module that is holding our testing utilities. But I'm very much trying to avoid making large file structure changes in this refactor.